dbms/src/Flash/Pipeline/Schedule/WaitReactor.cpp

```diff
@@ -122,35 +122,34 @@ void WaitReactor::waitForStop()
     LOG_INFO(logger, "wait reactor is stopped");
 }
 
-void WaitReactor::submit(TaskPtr && task) noexcept
+void WaitReactor::submit(TaskPtr && task)
 {
     waiting_task_list.submit(std::move(task));
 }
 
-void WaitReactor::submit(std::list<TaskPtr> & tasks) noexcept
+void WaitReactor::submit(std::list<TaskPtr> & tasks)
 {
     waiting_task_list.submit(tasks);
 }
 
-void WaitReactor::loop() noexcept
+void WaitReactor::loop()
+{
+    try
+    {
+        doLoop();
+    }
+    CATCH_AND_TERMINATE(logger)
+}
+
+void WaitReactor::doLoop()
 {
     setThreadName("WaitReactor");
     LOG_INFO(logger, "start wait reactor loop");
     ASSERT_MEMORY_TRACKER
 
     Spinner spinner{scheduler, logger};
     std::list<TaskPtr> local_waiting_tasks;
-    // Get the incremental tasks from waiting_task_list.
-    // return false if waiting_task_list has been closed.
-    auto take_from_waiting_task_list = [&]() {
-        return local_waiting_tasks.empty()
-            ? waiting_task_list.take(local_waiting_tasks)
-            // If the local waiting tasks are not empty, there is no need to be blocked here
-            // and we can continue to process the leftover tasks in the local waiting tasks
-            : waiting_task_list.tryTake(local_waiting_tasks);
-    };
-    while (take_from_waiting_task_list())
-    {
+    auto react = [&]() {
```

Comment:
Why ues lambda, IMHO this don't provide any readabilty.

Reply:
`react` has been called twice, do you mean to write it as a member function?

Reply:
> `react` has been called twice, do you mean to write it as a member function?

Yes